### PR TITLE
Add support for explicit FunctionRegistration

### PR DESCRIPTION
**Pull request by dsyer**
A bean of type FunctionRegistration registers the function with
user-specified name and other properties, rather than relying on the
bean name.
    
Alternatively, function catalog keys can be specified as a
@Qualifier, which will be used instead of the bean name if
no registration is found.

### CODE REVIEW

File: spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/FunctionRegistration.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2012-2015 the original author or authors.
```
Comment (markfisher) [inline spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/FunctionRegistration.java]: please update your template to 2017

File: spring-cloud-function-context/pom.xml [hunk 2]
```diff
@@ -25,6 +25,10 @@
 			<version>${spring-cloud-function.version}</version>
 		</dependency>
 		<dependency>
+			<groupId>org.springframework</groupId>
```
Comment (markfisher) [inline spring-cloud-function-context/pom.xml]: why is this needed?

File: spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+		if (context != null) {
+			context.close();
+		}
+	}
+
+	@Test
+	public void simpleFunction() {
+		create(SimpleConfiguration.class);
+		assertThat(context.getBean("function")).isInstanceOf(Function.class);
+		assertThat(catalog.lookupFunction("function")).isInstanceOf(Function.class);
```
Comment (markfisher) [inline spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java]: this doesn't actually test the `Function` itself (unlike the following 2 tests for `Supplier` and `Consumer`)

File: spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+		public Consumer<String> consumer() {
+			return value -> list.add(value);
+		}
+	}
+
+	@EnableAutoConfiguration
+	@Configuration
+	protected static class QualifiedConfiguration {
+		@Bean
+		@Qualifier("other")
```
Comment (markfisher) [inline spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java]: should we add an additional test for `@FunctionEntry`? (or on the other hand, do we want to hold off on introducing it and stick with plain `@Qualifier` for now?)

File: spring-cloud-function-context/pom.xml (hunk 2 above)
Comment (dsyer) [inline spring-cloud-function-context/pom.xml]: For Jackson. Otherwise there is no `ObjectMapper` configured in Spring Boot.

File: spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java (hunk 3 above)
Comment (dsyer) [inline spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java]: Is that a problem?

File: spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java (hunk 4 above)
Comment (dsyer) [inline spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java]: Yeah, I'm not sure on the name, so I left the test as `@Qualifier`. I think it would be good to have a custom annotation. Just not convinced on the name yet.

File: spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/FunctionRegistration.java (hunk 1 above)
Comment (dsyer) [inline spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/FunctionRegistration.java]: Will do. I assume this is not a blocker.

Comment (markfisher) [inline spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/FunctionRegistration.java]: will be updating all of these at some point, so not a blocker

File: spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java (hunk 3 above)
Comment (markfisher) [inline spring-cloud-function-context/test/java/org/springframework/cloud/function/context/ContextFunctionCatalogAutoConfigurationTests.java]: not really considering the primary purpose of this test... I had just noticed it wasn't consistent with the other tests below

File: spring-cloud-function-context/pom.xml (hunk 2 above)
Comment (markfisher) [inline spring-cloud-function-context/pom.xml]: ok, fine for now... at some point down the road, we should explore options for reducing dependencies in general

**Comment by markfisher** [comment]
rebased and merged as https://github.com/markfisher/spring-cloud-function/commit/d67159729da9277ba4a1b8a9cb06e150fa98349a
